lib/buffer.js

```diff
@@ -600,7 +601,21 @@ Buffer.concat = function concat(list, length) {
       throw new ERR_INVALID_ARG_TYPE(
         `list[${i}]`, ['Buffer', 'Uint8Array'], list[i]);
     }
-    pos += _copyActual(buf, buffer, pos, 0, buf.length);
+
+    if (buf.length === 0 || pos >= length) {
+      // Empty buf or destination is full; subsequent copies are no-ops.
+      // We don't break to preserve compatibility and throw if isUint8Array fails.
+      continue;
+    }
+
+    const available = length - pos;
+    const toCopy = buf.length > available ? available : buf.length;
+    if (toCopy === buf.length) {
```

Comment:
nit: we should probably avoid calling the getter so may times
```suggestion
    const bufLength = TypedArrayPrototypeGetLength(buf);
    if (bufLength === 0 || pos >= length) {
      // Empty buf or destination is full; subsequent copies are no-ops.
      // We don't break to preserve compatibility and throw if isUint8Array fails.
      continue;
    }

    const available = length - pos;
    const toCopy = bufLength > available ? available : bufLength;
    if (toCopy === bufLength) {
```

Reply:
Nice comment, I agree

However, I still think we shouldn't use `TypedArrayPrototypeGetLength`, it won't have the desired impact if the input is broken as broken input would break this part anyway, and we'd still over-allocate:

https://github.com/nodejs/node/blob/1502f5cdbaa434f86f900dc6e78a7011e7bacd8d/lib/buffer.js#L586-L589

If I use `TypedArrayPrototypeGetLength` there too, we will make a breaking change, because the validation is weirdly done later at:

https://github.com/nodejs/node/blob/1502f5cdbaa434f86f900dc6e78a7011e7bacd8d/lib/buffer.js#L599-L604

If someone passes something that's not a TypedArray, `TypedArrayPrototypeGetLength` would throw a different error before even our validation runs

Reply:
Ah, i missed this comment as outdated comments were hidden in gh ui (and new experimental ui is buggy / missed diffs lines on other PRs)

This smells of DoS. Non-blocking for this PR though as it doesn't change that